styx-scheduler-service/src/main/java/com/spotify/styx/state/QueuedStateManager.java

```diff
@@ -60,132 +59,203 @@
-  static final int EVENT_QUEUE_SIZE = 1024;
-  static final int POLL_TIMEOUT_MILLIS = 10;
-  static final int SHUTDOWN_GRACE_PERIOD_SECONDS = 5;
-  static final long NO_EVENTS_PROCESSED = -1L;
+  private static final long NO_EVENTS_PROCESSED = -1L;
+
+  private static final Duration SHUTDOWN_GRACE_PERIOD = Duration.ofSeconds(5);
 
   private final Time time;
-  private final Executor outputHandlerExecutor;
+
+  private final ExecutorService eventTransitionExecutor;
   private final Storage storage;
   private final BiConsumer<SequenceEvent, RunState> eventConsumer;
   private final Executor eventConsumerExecutor;
+  private final OutputHandler outputHandler;
 
-  private final ConcurrentMap<WorkflowInstance, InstanceState> states = Maps.newConcurrentMap();
-
-  private final Thread dispatcherThread;
-  private final CountDownLatch closedLatch = new CountDownLatch(1);
-  private final Object signal = new Object();
-  private AtomicInteger activeEvents = new AtomicInteger(0);
   private volatile boolean running = true;
 
   public QueuedStateManager(
       Time time,
-      Executor outputHandlerExecutor,
+      ExecutorService eventTransitionExecutor,
       Storage storage,
       BiConsumer<SequenceEvent, RunState> eventConsumer,
-      Executor eventConsumerExecutor) {
+      Executor eventConsumerExecutor,
+      OutputHandler outputHandler) {
     this.time = Objects.requireNonNull(time);
-    this.outputHandlerExecutor = Objects.requireNonNull(outputHandlerExecutor);
     this.storage = Objects.requireNonNull(storage);
     this.eventConsumer = Objects.requireNonNull(eventConsumer);
     this.eventConsumerExecutor = Objects.requireNonNull(eventConsumerExecutor);
-    this.dispatcherThread = new Thread(this::dispatch);
-    dispatcherThread.setName(DISPATCHER_THREAD_NAME);
-    dispatcherThread.start();
+    this.eventTransitionExecutor = Objects.requireNonNull(eventTransitionExecutor);
+    this.outputHandler = Objects.requireNonNull(outputHandler);
   }
 
   @Override
-  public void initialize(RunState runState) throws IsClosedException {
+  public CompletionStage<Void> trigger(WorkflowInstance workflowInstance, Trigger trigger)
+      throws IsClosedException {
     ensureRunning();
-
-    final WorkflowInstance workflowInstance = runState.workflowInstance();
-    if (states.containsKey(workflowInstance)) {
-      throw new AlreadyInitializedException("RunState initialization called on active instance "
-                                            + workflowInstance.toKey());
-    }
-
     final long counter;
     try {
       counter = storage.getLatestStoredCounter(workflowInstance).orElse(NO_EVENTS_PROCESSED);
     } catch (IOException e) {
-      throw Throwables.propagate(e);
+      return CompletableFutures.exceptionallyCompletedFuture(e);
     }
 
-    states.computeIfAbsent(workflowInstance, (wfi) -> new InstanceState(wfi, runState, counter + 1));
-  }
+    // TODO: optional retry on transaction conflict
```

Comment:
In what scenario would trigger lead to a conflict? I think this is a questionable TODO, the API caller only knows how long they'd like to retry.